src/qt/listctrl.cpp

```diff
@@ -13,125 +13,1020 @@
+
+        CopySelectStatusToItem(info, row, col);
+
+        return true;
+    }
+
+    bool IsVirtual() const wxOVERRIDE
+    {
+        return true;
+    }
+
+    void SetVirtualItemCount(long count) wxOVERRIDE
+    {
+        beginResetModel();
+        m_rowCount = static_cast<int>(count);
+        endResetModel();
+    }
+private:
+    int m_rowCount;
+};
+
+
+class wxQtListTreeWidget : public wxQtEventSignalHandler< QTreeView, wxListCtrl >
 {
 public:
     wxQtListTreeWidget( wxWindow *parent, wxListCtrl *handler );
 
-    void EmitListEvent(wxEventType typ, QTreeWidgetItem *qitem, int column) const;
+    void EmitListEvent(wxEventType typ, const QModelIndex &index) const;
 
-    void closeEditor(QWidget *editor, QAbstractItemDelegate::EndEditHint hint) wxOVERRIDE
+    void closeEditor(
+        QWidget *editor,
+        QAbstractItemDelegate::EndEditHint hint
+    ) wxOVERRIDE
     {
-        QTreeWidget::closeEditor(editor, hint);
-        m_editorFactory.ClearEditor();
+        // Close process can re-signal closeEditor so we need to guard against
+        // reentrant calls.
+        wxRecursionGuard guard(m_closingEditor);
+
+        if (guard.IsInside())
+            return;
+
+        // There can be multiple calls to close editor when the item loses focus
+        const QModelIndex current_index = m_itemDelegate.GetCurrentModelIndex();
+        if (!current_index.isValid())
+            return;
+
+        const wxString editedText = m_itemDelegate.GetEditControl()->GetLineText(0);
+
+        wxListEvent event;
+        InitListEvent(event,
+                      GetHandler(),
+                      wxEVT_LIST_END_LABEL_EDIT,
+                      current_index);
+        event.m_item.SetText(editedText);
+
+        if (hint == QAbstractItemDelegate::RevertModelCache)
+        {
+            event.SetEditCanceled(true);
+            EmitEvent(event);
+        }
+        else
+        {
+            // Allow event handlers to decide whether to accept edited text
+            if (!GetHandler()->HandleWindowEvent(event) || event.IsAllowed())
+                m_itemDelegate.AcceptModelData(editor, model(), current_index);
+        }
+
+        // wx doesn't have hints to edit next/previous item
+        if ( hint == QAbstractItemDelegate::EditNextItem ||
+             hint == QAbstractItemDelegate::EditPreviousItem )
+        {
+            hint = QAbstractItemDelegate::SubmitModelCache;
+        }
+
+        QTreeView::closeEditor(editor, hint);
+        closePersistentEditor(current_index);
     }
 
     wxTextCtrl *GetEditControl()
     {
-        return m_editorFactory.GetEditControl();
+        return m_itemDelegate.GetEditControl();
     }
 
-private:
-    void itemClicked(QTreeWidgetItem * item, int column);
-    void itemActivated(QTreeWidgetItem * item, int column);
-    void itemPressed(QTreeWidgetItem * item, int column);
-
-    void ChangeEditorFactory()
+    virtual void paintEvent(QPaintEvent *event) wxOVERRIDE
```

Comment:
Why do we need to override this? I saw that we get `QWidget::paintEngine: Should no longer be called` warnings if we don't, but it doesn't seem right to have to do it.

Reply:
The issue comes from `QTreeView`/`QTreeWidget` being composite controls (the header plus the contents).  I don't think QT intends the top level tree widget to be paintable.   When we wx handles QT's paint event (in `wxWindowQt::QtHandlePaintEvent`) it starts painting onto the `QTreeView`

Unfortunately,  I couldn't find a better workaround than suppressing wx's paint handling. Due to the way events are handled in QT via virtual functions,  "hiding" the wx paintEvent function is the only way I could prevent `wxWindowQt::QtHandlePaintEvent` for interfering with the tree widget. 

I do agree that it's a looks a bit strange and if anyone does have a better workaround then then I would definitely choose that over my solution.